runtime/src/accounts_index.rs

```diff
@@ -181,9 +211,201 @@ impl<'a, T: 'static + Clone> Iterator for AccountsIndexIterator<'a, T> {
     }
 }
 
+#[derive(Debug, Default)]
+struct SecondaryIndex {
+    // Map from index keys to index values
+    index: DashMap<Pubkey, SecondaryIndexEntry>,
+    // Map from index values back to index keys, used for cleanup.
+    // Alternative is to store Option<Pubkey> in each AccountInfo in the
+    // AccountsIndex if something is an SPL account with a mint, but then
+    // every AccountInfo would have to allocate `Option<Pubkey>`
+    reverse_index: DashMap<Pubkey, SecondaryReverseIndexEntry>,
+}
+
+impl SecondaryIndex {
+    fn insert(&self, key: &Pubkey, inner_key: &Pubkey, slot: Slot) {
+        {
+            let pubkeys_map = self
+                .index
+                .get(key)
+                .unwrap_or_else(|| self.index.entry(*key).or_insert(DashMap::new()).downgrade());
+
+            let slot_set = pubkeys_map.get(inner_key).unwrap_or_else(|| {
+                pubkeys_map
+                    .entry(*inner_key)
+                    .or_insert(RwLock::new(HashSet::new()))
+                    .downgrade()
+            });
+            let contains_key = slot_set.read().unwrap().contains(&slot);
+            if !contains_key {
+                slot_set.write().unwrap().insert(slot);
+            }
+        }
+
+        let prev_key = {
+            let slots_map = self.reverse_index.get(inner_key).unwrap_or_else(|| {
+                self.reverse_index
+                    .entry(*inner_key)
+                    .or_insert(RwLock::new(HashMap::new()))
+                    .downgrade()
+            });
+            let should_insert = {
+                // Most of the time, key should already exist and match
+                // the one in the update
+                if let Some(existing_key) = slots_map.read().unwrap().get(&slot) {
+                    existing_key != key
+                } else {
+                    // If there is no key yet, then insert
+                    true
+                }
+            };
+            if should_insert {
+                slots_map.write().unwrap().insert(slot, *key)
+            } else {
+                None
+            }
+        };
+
+        if let Some(prev_key) = prev_key {
+            // If the inner key was moved to a different primary key, remove
+            // the previous index entry.
+            if prev_key != *key {
```

Comment:
I think this check can be proofed to be unnecessary? i.e. if `prev_key`  is Some, it's always `prev_key` != `*key`.

Reply:
So the fundamental question is:

1) 
```
let should_insert = {
                // Most of the time, key should already exist and match
                // the one in the update
                if let Some(existing_key) = slots_map.read().unwrap().get(&slot) {
                    existing_key != key
                } else {
                    // If there is no key yet, then insert
                    true
                }
            };
```

2) 
```
if should_insert {
    slots_map.write().unwrap().insert(slot, *key)
}
```

Between the check in 1) and the write in 2) is it possible another thread already wrote the same `slot` to `slots_map`?

In the current implementation of AccountsDb, because a pubkey is exclusively updated by one thread per slot due to the Accounts locks, this may be true. But if ever in the future this changes, this may not hold. For instance if somebody was to try and introduce a bug like this: https://github.com/solana-labs/solana/pull/14212#discussion_r548774161 where you concurrently insert multiple versions of a pubkey from the *same* slot into the index at once 😉 

What do you think? Leave it out because it's currently unnecessary?

Reply:
ok, you're right. :)
let's retain the extra check adding some commentary like this:

```rust
// Interleaved other thread's writes between 
// `should_insert = { ... slots_map.get(...) ... }` and  `prev_key = { ... slots_map.insert(...) ... }`
// isn't possible due to current AccountsIndex's pubkey-per-thread exclusive-locking,
// but check here just in case for future more relaxed implementation
if prev_key != *key {
```

Reply:
added!